packages/firestore/src/remote/datastore.ts

```diff
@@ -225,11 +226,13 @@ export async function invokeRunQueryRpc(
   query: Query
 ): Promise<Document[]> {
   const datastoreImpl = debugCast(datastore, DatastoreImpl);
-  const request = toQueryTarget(datastoreImpl.serializer, queryToTarget(query));
+  const target = queryToTarget(query);
+  const request = toQueryTarget(datastoreImpl.serializer, target);
+  const encodedPath = toQueryTargetPath(datastoreImpl.serializer, target, true);
```

Comment:
This is the fundamental change in this PR. We get an encoded query path from the new method `gotQueryTargetPath(...)`, instead of using the unencoded query path from `toQueryTarget(...).parent`.

The difference between the unencoded query target path and encoded path is like: 
"projects/myproject/databases/(default)/documents/doc+with+plus/collection+with+plus/" <-- unencoded
"projects/myproject/databases/(default)/documents/doc%2bwith%2bplus/collection%2bwith%2bplus/" <-- encoded

This encoding is achieved by sending path segments through the system [`encodeUriComponent(...)`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/encodeURIComponent)

Reply:
All other code changes here are refactoring to make appropriate re-use of a somewhat deep call stack that converts a target to a path.